lib/ecto/repo/schema.ex

```diff
@@ -76,6 +86,43 @@ defmodule Ecto.Repo.Schema do
     end
   end
 
+  defp extract_header_and_fields(query = %Ecto.Query{}, _schema, _dumper, _autogen_id, _placeholder_map, adapter) do
+    case query.select do
+      %Ecto.Query.SelectExpr{expr: {:%{}, _ctx, args}} ->
+        header =
+          Keyword.keys(args)
+          |> Enum.reduce(%{}, &Map.put(&2, &1, true))
+
+        query = Map.update!(query, :select, fn select ->
+          Map.update!(select, :expr, fn {op, ctx, args} ->
+            args = Enum.sort_by(args, &elem(&1, 0))
+            {op, ctx, args}
+          end)
+        end)
+
+        query_and_params = Ecto.Adapter.Queryable.plan_query(:all, adapter, query)
+
+        {query_and_params, header, []}
+
+      _ ->
+        raise ArgumentError, message: """
+        Cannot generate a fields list for insert_all from the given source query,
+        because it does not have a select clause that uses a map:
+        #{inspect query}
+
+        Please add a select clause that selects into a map, like this:
+
+        from x in Source,
+          ...,
+          select: %{
+            field_a: x.bar,
+            field_b: x.foo
+          }
+
+        The keys must exist in the schema that is being inserted into.
+        """
```

Comment:
```suggestion
        raise ArgumentError, """
        Cannot generate a fields list for insert_all from the given source query
        because it does not have a select clause that uses a map:

        #{inspect query}

        Please add a select clause that selects into a map, like this:

            from x in Source,
              ...,
              select: %{
                field_a: x.bar,
                field_b: x.foo
              }

        The keys must exist in the schema that is being inserted into
        """
```